lib/src/modules/task.cc

```diff
@@ -128,11 +149,145 @@ static TaskCommand getTaskCommand(const fs::path &task_executable)
     return TaskCommand { task_executable.string(), { } };
 }
 
+// Converts boost filesystem errors to a task_error object.
+static Module::ProcessingError toModuleProcessingError(const fs::filesystem_error& e) {
+    std::string err_msg = e.what();
+    if (alg::contains(err_msg, "No such file or directory")) {
+      return Module::ProcessingError(lth_loc::format("No such file or directory: {1}", e.path1()));
+    } else if (alg::contains(err_msg, "File exists")) {
+      return Module::ProcessingError(lth_loc::format("A file matching the name of the provided sha already exists"));
+    }
+    return Module::ProcessingError(err_msg);
+}
+
+// Take out common permissions.
+static fs::perms TASK_FILE_PERMS = fs::owner_read | fs::owner_write | fs::group_read;
+
+// Creates the <task_cache_dir>/<sha256> directory for a single
+// task, ensuring that its permissions are readable by
+// the PXP agent owner/group (for unix OSes), writable for the PXP agent owner,
+// and executable by both PXP agent owner and group. Returns the path to this directory.
+// Note that the last modified time of the directory is updated, and that this routine
+// will not fail if the directory already exists.
+static fs::path createCacheDir(const fs::path& task_cache_dir, const std::string& sha256) {
+    auto cache_dir = fs::path(task_cache_dir) / sha256;
+    fs::create_directory(cache_dir);
+    fs::last_write_time(cache_dir, time(nullptr));
+    fs::permissions(cache_dir, TASK_FILE_PERMS | fs::owner_exe | fs::group_exe);
+    return cache_dir;
+}
+
+// Computes the sha256 of the file denoted by path. Assumes that
+// the file designated by "path" exists.
+static std::string calculateSha256(const std::string& path) {
+    const std::streamsize CHUNK_SIZE = 2 << 14;
+    char buffer[CHUNK_SIZE];
+    unsigned char md_value[EVP_MAX_MD_SIZE];
+    unsigned int md_len;
+
+    auto md = EVP_sha256();
+    auto mdctx = EVP_MD_CTX_create();
+    EVP_DigestInit_ex(mdctx, md, nullptr);
+    boost::nowide::ifstream ifs(path);
+    while (ifs.read(buffer, CHUNK_SIZE)) {
+      EVP_DigestUpdate(mdctx, buffer, CHUNK_SIZE);
```

Comment:
The `CHUNK_SIZE` needs to be replaced with [`ifs.gcount()`](http://en.cppreference.com/w/cpp/io/basic_istream/gcount) otherwise if the last chunk is not exactly CHUNK_SIZE long the checksum computation will process garbage (actually the end of the las but one chunk) at the end of the buffer.